### PR TITLE
rustbuild: A few tweaks

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes https://github.com/rust-lang/rust/issues/40016
Fixes https://github.com/rust-lang/rust/issues/39507

r? @alexcrichton

### CODE REVIEW

File: src/build_helper/lib.rs [hunk 1]
```diff
@@ -148,6 +148,41 @@ pub fn up_to_date(src: &Path, dst: &Path) -> bool {
...
 }
 
+pub struct NativeLibBoilerplate {
+    pub skip_build: bool,
+    pub src_dir: PathBuf,
+    pub out_dir: PathBuf,
+    pub timestamp: PathBuf,
+}
+
+pub fn native_lib_boilerplate(src_name: &str,
```
Comment (alexcrichton) [inline src/build_helper/lib.rs]: Could you add some comments here as well as to how this function is expected to be called? It looks like there's a standard rigamarole of checking `skip_build` and then creating the timestamp at the end, and it'd be good to enusre that's all documented.

**Comment by alexcrichton** [comment]
Just one minor request for a comment but otherwise looks great to me, thanks!

r=me

**Comment by petrochenkov** [comment]
@alexcrichton 
Updated with comments for `native_lib_boilerplate`.
I also automated timestamp creation and partially build skipping, it should be simpler this way.

**Comment by petrochenkov** [comment]
Spurious (?) failure in the middle of LLVM build due to sccache (?).
Or it may be some interaction between LLVM build caching used on CI and https://github.com/rust-lang/rust/pull/40236/commits/de2af63b469d77bdadcbec51baf117efe0d88fc0.

@bors retry

**Comment by petrochenkov** [comment]
@alexcrichton 

What are requirement for building LLVM on Travis/Appveyor with caching, by the way?

What I want with local builds is:
- if `llvm-finished-building` does not exist, do not delete the build directory, reuse it (**REUSE**)
- if `llvm-finished-building` exists and differs from `llvm-auto-clean-trigger`, do not delete the build directory, reuse it (**REUSE**)
- if `llvm-finished-building` exists and is equal to `llvm-auto-clean-trigger`, the do not try to build anything at all (**SKIP**)
- fresh build is done only if something really bad happens and I delete the build directory manually, so this PR doesn't quite do what I'd like to see ideally.

As I understand, what CI wants is:
- if `llvm-finished-building` does not exist, delete the build directory, do a fresh build accelerated by (s)ccache (**FRESH**)
- if `llvm-finished-building` exists and differs from `llvm-auto-clean-trigger`, delete the build directory, do a fresh build accelerated by (s)ccache (**FRESH**)
- if `llvm-finished-building` exists and is equal to `llvm-auto-clean-trigger`, the do not try to build anything at all (**SKIP**)
- i.e. CI always does either full reuse or, failing that, accelerated fresh build, because reuse provided by caching tools is considered more reliable than one provided by CMake.

Is my understanding correct?
If it's correct, then a new option in `config.toml` switching between these use cases would be a better solution than what https://github.com/rust-lang/rust/pull/40236/commits/de2af63b469d77bdadcbec51baf117efe0d88fc0 does.

**Pull request merged**